Assets/SEE/Dissonance/ChatLogController.cs

```diff
@@ -0,0 +1,327 @@
+            // Directly set the height to that height to ensure all text is seen.
+            txt.GetComponent<RectTransform>().sizeDelta = new Vector2(0, txt.preferredHeight);
+            txt.GetComponent<RectTransform>().anchoredPosition += new Vector2(0, 3);
+
+            // Save the item in the queue of log entries.
+            logEntries.Enqueue(new ChatLogEntry(txt));
+
+            // Bump all items up by the appropriate amount.
+            foreach (var chatLogEntry in logEntries)
+            {
+                chatLogEntry.Transform.anchoredPosition += new Vector2(0, txt.preferredHeight);
+
+                // If this item is over the height limit fade it out and remove it from the log when faded out.
+                if (chatLogEntry.Transform.anchoredPosition.y > heightLimit && !chatLogEntry.IsTransitioningOut)
+                {
+                    chatLogEntry.FadeOut();
+                }
+            }
+
+            // Make sure the UI is visible for some time.
+            ShowFor(TimeSpan.FromSeconds(3));
+        }
+
+        /// <summary>
+        /// Fades out the chat log if the time has come and removes entries that
+        /// are totally faded out.
+        /// </summary>
+        public void Update()
+        {
+            if (ForceShow)
+            {
+                // A new cycle of showing the chat log has started.
+                canvasGroup.alpha = 1;
+                fadeOutStartTime = DateTime.UtcNow + TimeSpan.FromSeconds(4);
+            }
+            else if (fadeOutStartTime < DateTime.UtcNow)
+            {
+                // If now is beyond fadeOutStartTime, fade out the chat log.
+                float time = (float)(DateTime.UtcNow - fadeOutStartTime).TotalSeconds / 2;
+                if (time > 0)
+                {
+                    float alpha = (2 - Mathf.Clamp(time, 0, 2)) / 2;
+                    canvasGroup.alpha = alpha;
+                    // If the chat log is almost invisible, disable the canvas altogether.
+                    if (alpha <= 0.05)
+                    {
+                        EnableCanvas(false);
+                    }
+                }
+            }
+            else if (fadeOutStartTime > DateTime.UtcNow)
+            {
+                canvasGroup.alpha = 1;
+            }
+
+            // Remove entries which no longer fit into the chat log and are fully faded out.
+            while (logEntries.Count > 0 && logEntries.Peek().IsTransitionComplete)
+            {
+                Destroy(logEntries.Dequeue().Object);
+            }
+
+            // Update animations on fading chat lines.
+            foreach (ChatLogEntry chatLogEntry in logEntries)
+            {
+                if (chatLogEntry.IsTransitioningOut)
+                {
+                    chatLogEntry.Update();
+                }
+                else
+                {
+                    break;
+                }
```

Comment:
Can be simplified (suggestion may need LINQ import to work):
```suggestion
            foreach (ChatLogEntry chatLogEntry in logEntries.TakeWhile(x => x.IsTransitioningOut))
            {
                chatLogEntry.Update();
```